### PR TITLE
Upgrade to Vert.x 4.0.0 MODAT-96

### PR DESCRIPTION
**Pull request by adamdickmeiss**
1. commit is the bare minimum to get tests to pass.
2. makes it use log4j2.
3. Futurisation
4. Update README

### CODE REVIEW

File: src/main/java/org/folio/auth/authtokenmodule/impl/ModulePermissionsSource.java [hunk 1]
```diff
@@ -73,70 +75,62 @@ public static void setCacheTimeout(int sec) {
...
-            if (permissionsObject != null && permissionsObject.getJsonArray("permissionNames") != null) {
-              logger.debug("Got permissions");
-              promise.complete(permissionsObject.getJsonArray("permissionNames"));
-            } else {
-              logger.error("Got malformed/empty permissions object");
-              promise.fail("Got malformed/empty permissions object");
-            }
-          });
+          final String requestUrl = okapiUrl + "/perms/users/" + permUser.getString("id") + "/permissions?expanded=true";
+          logger.debug("Requesting permissions from URL at " + requestUrl);
```
Comment (julianladisch) [inline src/main/java/org/folio/auth/authtokenmodule/impl/ModulePermissionsSource.java:99]: Should use {} placeholder.

File: src/main/java/org/folio/auth/authtokenmodule/impl/ModulePermissionsSource.java [hunk 2]
```diff
@@ -73,70 +75,62 @@ public static void setCacheTimeout(int sec) {
...
-        promise.fail(e);
-      });
-    });
-    permUserReq.exceptionHandler(promise::fail);
-    endRequest(permUserReq, requestToken, tenant, requestId);
-    return promise.future();
   }
 
-  private void endRequest(HttpClientRequest req, String requestToken,
+  private void endRequest(HttpRequest<Buffer> req, String requestToken,
```
Comment (julianladisch) [inline src/main/java/org/folio/auth/authtokenmodule/impl/ModulePermissionsSource.java:133]: endRequest should be renamed, for example setHeaders

File: src/main/java/org/folio/auth/authtokenmodule/impl/ModulePermissionsSource.java [hunk 3]
```diff
@@ -175,55 +168,35 @@ private void endRequest(HttpClientRequest req, String requestToken,
...
-      });
-      endRequest(req, requestToken, tenant, requestId);
-    } catch (Exception e) {
-      logger.error(e.getLocalizedMessage(), e);
-      promise.fail("Unable to expand permissions: " + e.getLocalizedMessage());
-    }
-    return promise.future();
+    String requestUrl = okapiUrl + "/perms/permissions?"
+        + "expanded=true&query=" + URLEncoder.encode(query, StandardCharsets.UTF_8);
+    logger.debug("Requesting expanded permissions from URL at " + requestUrl);
```
Comment (julianladisch) [inline src/main/java/org/folio/auth/authtokenmodule/impl/ModulePermissionsSource.java:174]: Should use a placeholder {}

Comment (adamdickmeiss) [reply src/main/java/org/folio/auth/authtokenmodule/impl/ModulePermissionsSource.java]: yep

File: src/main/java/org/folio/auth/authtokenmodule/impl/ModulePermissionsSource.java (hunk 2 above)
Comment (adamdickmeiss) [reply src/main/java/org/folio/auth/authtokenmodule/impl/ModulePermissionsSource.java]: Makes sense.

File: src/main/java/org/folio/auth/authtokenmodule/impl/ModulePermissionsSource.java (hunk 1 above)
Comment (adamdickmeiss) [reply src/main/java/org/folio/auth/authtokenmodule/impl/ModulePermissionsSource.java]: yep

File: src/main/java/org/folio/auth/authtokenmodule/impl/ModulePermissionsSource.java [hunk 4]
```diff
@@ -73,71 +75,63 @@ public static void setCacheTimeout(int sec) {
...
-      });
-    });
-    permUserReq.exceptionHandler(promise::fail);
-    endRequest(permUserReq, requestToken, tenant, requestId);
-    return promise.future();
   }
 
-  private void endRequest(HttpClientRequest req, String requestToken,
-                          String tenant, String requestId) {
+  private void sendHeaders(HttpRequest<Buffer> req, String requestToken,
```
Comment (julianladisch) [inline src/main/java/org/folio/auth/authtokenmodule/impl/ModulePermissionsSource.java:133]: sendHeaders should be setHeaders

**Review by julianladisch — APPROVED** [review]
Looks good, thanks!

File: src/main/java/org/folio/auth/authtokenmodule/impl/ModulePermissionsSource.java (hunk 4 above)
Comment (adamdickmeiss) [reply src/main/java/org/folio/auth/authtokenmodule/impl/ModulePermissionsSource.java]: ok

**Pull request merged**